PWGJE/TableProducer/emcalCorrectionTask.cxx

```diff
@@ -174,238 +178,262 @@ struct EmcalCorrectionTask {
   // void process(aod::Collision const& collision, aod::Tracks const& tracks, aod::Calos const& cells)
   // void process(aod::BCs const& bcs, aod::Collision const& collision, aod::Calos const& cells)
 
+  Preslice<collEventSels> collisionsPerFoundBC = aod::evsel::foundBCId;
+  // Use Preslice once binding node is available
+  // Preslice<filteredCells> calosPerFoundBC = aod::calo::BCId;
+
   //  Appears to need the BC to be accessed to be available in the collision table...
-  void process(aod::BC const& bc, o2::soa::Join<aod::Collisions, aod::EvSels> const& collisions, myGlobTracks const& tracks, aod::Calos const& cells)
+  void process(bcEvSels const& bcs, collEventSels const& collisions, myGlobTracks const& tracks, filteredCells const& cells)
   {
     LOG(debug) << "Starting process.";
-    // Convert aod::Calo to o2::emcal::Cell which can be used with the clusterizer.
-    // In particular, we need to filter only EMCAL cells.
-    mEmcalCells.clear();
-    mCellIdToCellGlobalIndex.clear();
-    int c = 0;
-    hBC->Fill(7);
-    bool hasEMCCell = false;
+
+    // Preprocess cells
+    // Sort them according to BC indices, later made use of in the loop over BCs
+    // Struct must be a pair of vector of cells - vector of global indices of cells with 1-1 correspondence of vector
+    // indices between cell and cellindex
+    std::map<int, std::pair<std::shared_ptr<std::vector<o2::emcal::Cell>>, std::shared_ptr<std::vector<int64_t>>>> cellContainer;
     for (auto& cell : cells) {
-      if (cell.caloType() != selectedCellType) {
-        LOG(debug) << "Rejected";
-        continue;
+      auto bcGlobal = cell.bc_as<bcEvSels>().globalBC();
+      auto foundBC = cellContainer.find(bcGlobal);
+      std::shared_ptr<std::vector<o2::emcal::Cell>> cellsBC;
+      std::shared_ptr<std::vector<int64_t>> cellIndicesBC;
+      if (foundBC != cellContainer.end()) {
+        cellsBC = foundBC->second.first;
+        cellIndicesBC = foundBC->second.second;
+      } else {
+        cellsBC = std::make_shared<std::vector<o2::emcal::Cell>>();
+        cellIndicesBC = std::make_shared<std::vector<int64_t>>();
+        cellContainer[bcGlobal] = {cellsBC, cellIndicesBC};
```

Comment:
Maybe this can be optimised by using try_emplace and reserve:
```c++
std::map<int, std::pair<std::shared_ptr<std::vector<o2::emcal::Cell>>, std::shared_ptr<std::vector<int64_t>>>> cellContainer;
for (auto& cell : cells) {
  auto bcGlobal = cell.bc_as<bcEvSels>().globalBC();
  auto [cellsBC, cellIndicesBC] = cellContainer.try_emplace(bcGlobal, std::make_shared<std::vector<o2::emcal::Cell>>(), std::make_shared<std::vector<int64_t>>())->second;
  cellsBC->reserve(cellsBC->size() + 1);
  cellIndicesBC->reserve(cellIndicesBC->size() + 1);
  cellsBC->emplace_back(cell.cellNumber(),
                        cell.amplitude(),
                        cell.time(),
                        o2::emcal::intToChannelType(cell.cellType()));
  cellIndicesBC->emplace_back(cell.globalIndex());
}
```

Reply:
The try_emplace syntax looks indeed nice (apart from that value must be std::pair I think (two containers). The reserve not sure what it is needed for, emplace_back should enlarge the container in any case.

Reply:
I am not sure about the evaluation order. If make_shared is indeed evaluated before you are right, this would be an expensice memory allocation that we don't need that often.
The reserve with size +1 should reduce the number of memory allocations that would be needed if we just use emplace_back, if I remember correctly.